clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1308,14 +1326,51 @@ class SyclKernelBodyCreator
       DRE = ImplicitCastExpr::Create(SemaRef.Context, FD->getType(),
                                      CK_AddressSpaceConversion, DRE, nullptr,
                                      VK_RValue);
+    return DRE;
+  }
+
+  void createExprForStructOrScalar(FieldDecl *FD) {
+    InitializedEntity Entity =
+        InitializedEntity::InitializeMember(FD, &VarEntity);
     InitializationKind InitKind =
         InitializationKind::CreateCopy(SourceLocation(), SourceLocation());
+    Expr *DRE = createInitExpr(FD);
     InitializationSequence InitSeq(SemaRef, Entity, InitKind, DRE);
-
     ExprResult MemberInit = InitSeq.Perform(SemaRef, Entity, InitKind, DRE);
     InitExprs.push_back(MemberInit.get());
   }
 
+  void createExprForScalarElement(FieldDecl *FD, QualType FieldTy) {
+    InitializedEntity ArrayEntity =
+        InitializedEntity::InitializeMember(FD, &VarEntity);
+    InitializationKind InitKind =
+        InitializationKind::CreateCopy(SourceLocation(), SourceLocation());
+    Expr *DRE = createInitExpr(FD);
+    Expr *Idx = dyn_cast<ArraySubscriptExpr>(MemberExprBases.back())->getIdx();
+    llvm::APSInt Result;
+    SemaRef.VerifyIntegerConstantExpression(Idx, &Result);
+    uint64_t IntIdx = Result.getZExtValue();
+    InitializedEntity Entity = InitializedEntity::InitializeElement(
+        SemaRef.getASTContext(), IntIdx, ArrayEntity);
+    InitializationSequence InitSeq(SemaRef, Entity, InitKind, DRE);
+    ExprResult MemberInit = InitSeq.Perform(SemaRef, Entity, InitKind, DRE);
+    llvm::SmallVector<Expr *, 16> ArrayInitExprs;
+    if (IntIdx > 0) {
+      // Continue with the current InitList
+      InitListExpr *ILE = cast<InitListExpr>(InitExprs.back());
+      InitExprs.pop_back();
+      llvm::ArrayRef<Expr *> L = ILE->inits();
+      for (size_t I = 0; I < L.size(); I++)
+        ArrayInitExprs.push_back(L[I]);
+    }
```

Comment:
I feel that it can look simpler if we use approach like Elizabeth used for struct initializing here - https://github.com/intel/llvm/pull/1877/files#diff-c88a590aca6d5f8e92dd1f8b6ec05a49R1600 . When we are inside array, we just push initializers for array elements, when we leave array, we pop out initializers for array elements and create initializer list for the whole array.
Is it possible here?

Reply:
Done.